File: sdk/client-ts/src/clients/relayer.ts
import {
  Address,
  TransactionReceipt,
  TypedDataDefinition,
  formatEther,
} from 'viem';
import {
  Relayer,
  Transaction,
  TransactionSent,
  TransactionStatus,
  TransactionStatusResult,
  TransactionToSend,
  cancelTransaction,
  getRelayer,
  getRelayerAllowlistAddress,
  getTransaction,
  getTransactionStatus,
  getTransactions,
  replaceTransaction,
  sendTransaction,
} from '../api';
import {
  ApiBaseConfig,
  PagingContext,
  PagingResult,
  defaultPagingContext,
} from '../api/types';
import { Provider } from '../provider';
import { invariant } from '../utils';
import {
  signText,
  SignTextResult,
  signTypedData,
  SignTypedDataResult,
} from '../api';

export interface RelayerClientConfig {
  serverUrl: string;
  providerUrl: string;
  relayerId: string;
  auth:
    | {
        apiKey: string;
      }
    | {
        username: string;
        password: string;
      };
}

export class RelayerClient {
  public readonly id: string;
  protected readonly _apiBaseConfig: ApiBaseConfig;
  private readonly _ethereumProvider: Provider;
  constructor(config: RelayerClientConfig) {
    this.id = config.relayerId;
    this._ethereumProvider = new Provider(config.providerUrl, this);
    if ('apiKey' in config.auth) {
      this._apiBaseConfig = {
        serverUrl: config.serverUrl,
        apiKey: config.auth.apiKey,
      };
    } else {
      this._apiBaseConfig = {
        serverUrl: config.serverUrl,
        username: config.auth.username,
        password: config.auth.password,
      };
    }
  }

  /**
   * Get the relayer address
   * @returns string
   */
  public async address(): Promise<Address> {
    return (await this.getInfo()).address;
  }

  /**
   * Get the relayer information
   * @returns Relayer
   */
  public async getInfo(): Promise<Relayer> {
    const result = await getRelayer(this.id, this._apiBaseConfig);

    invariant(result, 'Relayer not found');

    return result.relayer;
  }

  /**
   * Get the relayer balance
   * @returns string
   */
  public async getBalanceOf(): Promise<string> {
    // @ts-ignore - use the viem getBalance method without exposing the client (which is not needed here)
    const balance = await this._ethereumProvider._client.getBalance({
      address: await this.address(),
    });

    return formatEther(balance);
  }

  public get allowlist() {
    return {
      /**
       * Get the relayer allowlist
       * @returns An address of allowlist addresses
       */
      get: (
        pagingContext: PagingContext = defaultPagingContext
      ): Promise<PagingResult<string>> => {
        return getRelayerAllowlistAddress(
          this.id,
          pagingContext,
          this._apiBaseConfig
        );
      },
    };
  }

  public get sign() {
    return {
      /**
       * Sign a message
       * @param message The message to sign
       * @param rateLimitKey Optional rate limit key
       * @returns SignTextResult
       */
      text: (
        message: string,
        rateLimitKey?: string
      ): Promise<SignTextResult> => {
        return signText(this.id, message, rateLimitKey, this._apiBaseConfig);
      },
      /**
       * Sign typed data
       * @param typedData The typed data to sign
       * @param rateLimitKey Optional rate limit key
       * @returns SignTypedDataResult
       */
      typedData: (
        typedData: TypedDataDefinition,
        rateLimitKey?: string
      ): Promise<SignTypedDataResult> => {
        return signTypedData(
          this.id,
          typedData,
          rateLimitKey,
          this._apiBaseConfig
        );
      },
    };
  }

  public get transaction() {
    return {
      /**
       * Get a transaction
       * @param transactionId The transaction id
       * @returns Transaction | null
       */
      get: (transactionId: string): Promise<Transaction | null> => {
        return getTransaction(transactionId, this._apiBaseConfig);
      },
      /**
       * Get a transaction status
       * @param transactionId The transaction id
       * @returns TransactionStatusResult | null
       */
      getStatus: (
        transactionId: string
      ): Promise<TransactionStatusResult | null> => {
        return getTransactionStatus(transactionId, this._apiBaseConfig);
      },
      /**
       * Get transactions for relayer
       * @returns Transaction[]
       */
      getAll: (
        pagingContext: PagingContext = defaultPagingContext
      ): Promise<PagingResult<Transaction>> => {
        return getTransactions(this.id, pagingContext, this._apiBaseConfig);
      },
      /**
       * Replace a transaction
       * @param transactionId The transaction id
       * @param replacementTransaction The replacement transaction
       * @param rateLimitKey The rate limit key if you want rate limit feature on
       * @returns transactionId
       */
      replace: (
        transactionId: string,
        replacementTransaction: TransactionToSend,
        rateLimitKey?: string | undefined
      ): Promise<TransactionSent> => {
        return replaceTransaction(
          transactionId,
          replacementTransaction,
          rateLimitKey,
          this._apiBaseConfig
        );
      },
      /**
       * cancel a transaction
       * @param transactionId The transaction id
       * @param rateLimitKey The rate limit key if you want rate limit feature on
       * @returns boolean
       */
      cancel: (
        transactionId: string,
        rateLimitKey?: string | undefined
      ): Promise<boolean> => {
        return cancelTransaction(
          transactionId,
          rateLimitKey,
          this._apiBaseConfig
        );
      },
      /**
       * Send a transaction
       * @param transaction The transaction to send
       * @param rateLimitKey The rate limit key if you want rate limit feature on
       * @returns transactionId
       */
      send: (
          transaction: TransactionToSend,
          rateLimitKey?: string | undefined
      ): Promise<TransactionSent> => {
        return sendTransaction(
          this.id,
          transaction,
          rateLimitKey,
          this._apiBaseConfig
        );
      },
      waitForTransactionReceiptById: async (
        transactionId: string
      ): Promise<TransactionReceipt> => {
        while (true) {
          const result = await this.transaction.getStatus(transactionId);
          if (!result) {
            throw new Error('Transaction not found');
          }

          switch (result.status) {
            case TransactionStatus.PENDING:
            case TransactionStatus.INMEMPOOL:
              await new Promise((resolve) => setTimeout(resolve, 500));
              break;
            case TransactionStatus.MINED:
            case TransactionStatus.CONFIRMED:
            case TransactionStatus.FAILED:
              invariant(result.receipt, 'Transaction receipt not found');
              return result.receipt;
            case TransactionStatus.EXPIRED:
              throw new Error('Transaction expired');
            default:
              throw new Error('Unknown transaction status');
          }
        }
      },
    };
  }

  public ethereumProvider(): Provider {
    return this._ethereumProvider;
  }
}

File: sdk/python/src/rrelayer/relayer.py
from pydantic import BaseModel, ConfigDict, PrivateAttr, validate_call
from web3 import AsyncWeb3, Web3
from typing import Any
from rrelayer.api import API
from rrelayer.types import PagingContext, defaultPagingContext


class RelayerClient(BaseModel):
    class AllowList:
        def __init__(self, relayerClient: "RelayerClient"):
            self._relayer: "RelayerClient" = relayerClient

        @validate_call
        async def get(self, pagingContext: PagingContext = defaultPagingContext):
            params = pagingContext.model_dump()

            return await self._relayer._api.getApi(
                self._relayer._apiBaseConfig,
                f"relayers/{self._relayer._id}/allowlists",
                params,
            )

    class Sign:
        def __init__(self, relayerClient: "RelayerClient"):
            self._relayer: "RelayerClient" = relayerClient

        @validate_call
        async def text(self, message: str, rateLimitKey: str = ""):
            pass

        @validate_call
        async def textHistory(
            self, pagingContext: PagingContext = defaultPagingContext
        ):
            pass

    class Transaction:
        def __init__(self, relayerClient: "RelayerClient"):
            self._relayer: "RelayerClient" = relayerClient

        async def get(self):
            pass

        async def getStatus(self):
            pass

        async def getAll(self):
            pass

        async def replace(self):
            pass

        async def cancel(self):
            pass

        async def send(self):
            pass

    _id: str = PrivateAttr()

    _apiBaseConfig: dict[str, str] = PrivateAttr()
    _ethereumProvider: AsyncWeb3 = PrivateAttr()

    _api: API = PrivateAttr()

    _allowlist: AllowList = PrivateAttr()
    _sign: Sign = PrivateAttr()
    _transaction: Transaction = PrivateAttr()

    model_config = ConfigDict(arbitrary_types_allowed=True)

    def __init__(
        self,
        serverURL: str,
        providerUrl: str,
        relayerId: str,
        auth: dict[str, str],
        **data,
    ):
        super().__init__(**data)

        self._id = relayerId

        self._ethereumProvider = AsyncWeb3(AsyncWeb3.AsyncHTTPProvider(providerUrl))

        if "apiKey" in auth:
            self._apiBaseConfig = {"apiKey": auth["apiKey"], "serverURL": serverURL}
        elif "username" in auth and "password" in auth:
            self._apiBaseConfig = {
                "username": auth["username"],
                "password": auth["password"],
                "serverURL": serverURL,
            }
        else:
            raise ValueError("Invalid authentication credentials")

        self._api = API()

        self._allowlist = self.AllowList(self)
        self._sign = self.Sign(self)
        self._transaction = self.Transaction(self)

    @property
    def name(self):
        return self._id

    @property
    def allowlist(self) -> AllowList:
        return self._allowlist

    @property
    def sign(self) -> Sign:
        return self._sign

    @property
    def transaction(self) -> Transaction:
        return self._transaction

    async def address(self) -> str | None:
        response = await self.getInfo()
        return Web3.to_checksum_address(response["address"]) if response else None

    async def getInfo(self):
        response = await self._api.getApi(self._apiBaseConfig, f"relayers/{self._id}")
        return response["relayer"] if response else None

    async def getBalanceOf(self):
        address = await self.address()
        if address:
            balance = await self._ethereumProvider.eth.get_balance(
                Web3.to_checksum_address(address)
            )
            return Web3.from_wei(balance, "ether")
        else:
            return 0

    def ethereumProvider(self) -> AsyncWeb3[Any]:
        return self._ethereumProvider
